File: src/CSharp/CodeCracker/Style/InterfaceNameAnalyzer.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;
using System.Collections.Immutable;

namespace CodeCracker.CSharp.Style
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class InterfaceNameAnalyzer : DiagnosticAnalyzer
    {
        internal const string Title = "You should add letter 'I' before interface name.";
        internal const string MessageFormat = "Consider naming interfaces starting with 'I'.";
        internal const string Category = SupportedCategories.Style;
        const string Description = "Consider naming interfaces starting with 'I'.";

        internal static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
            DiagnosticId.InterfaceName.ToDiagnosticId(),
            Title,
            MessageFormat,
            Category,
            DiagnosticSeverity.Info,
            isEnabledByDefault: true,
            description: Description,
            helpLinkUri: HelpLink.ForDiagnostic(DiagnosticId.InterfaceName));

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);

        public override void Initialize(AnalysisContext context) => context.RegisterSyntaxNodeAction(AnalyzeClass, SyntaxKind.InterfaceDeclaration);

        private static void AnalyzeClass(SyntaxNodeAnalysisContext context)
        {
            if (context.IsGenerated()) return;
            var invocationExpression = (InterfaceDeclarationSyntax)context.Node;
            var semanticModel = context.SemanticModel;

            var name = invocationExpression.Identifier.ToString().ToUpper();
            if (name.StartsWith("I")) return;
            var error = string.Format(MessageFormat, MessageFormat);
            var diag = Diagnostic.Create(Rule, invocationExpression.GetLocation(), error);
            context.ReportDiagnostic(diag);
        }
    }
}

File: src/VisualBasic/CodeCracker/Style/InterfaceNameAnalyzer.vb
Imports System.Collections.Immutable
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Diagnostics
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Style
    <DiagnosticAnalyzer(LanguageNames.VisualBasic)>
    Public Class InterfaceNameAnalyzer
        Inherits DiagnosticAnalyzer

        Friend Const Title = "You should add letter 'I' before interface name."
        Friend Const MessageFormat = "Consider naming interfaces starting with 'I'"
        Friend Const Category = SupportedCategories.Style
        Private Const Description = "Consider naming interfaces starting with 'I'."

        Friend Shared Rule As New DiagnosticDescriptor(
            DiagnosticId.InterfaceName.ToDiagnosticId(),
            Title,
            MessageFormat,
            Category,
            DiagnosticSeverity.Info,
            isEnabledByDefault:=True,
            description:=Description,
            helpLinkUri:=HelpLink.ForDiagnostic(DiagnosticId.InterfaceName)
        )

        Public Overrides ReadOnly Property SupportedDiagnostics As ImmutableArray(Of DiagnosticDescriptor)
            Get
                Return ImmutableArray.Create(Rule)
            End Get
        End Property

        Public Overrides Sub Initialize(context As AnalysisContext)
            context.RegisterSyntaxNodeAction(AddressOf AnalyzeClass, SyntaxKind.InterfaceStatement)
        End Sub

        Private Shared Sub AnalyzeClass(context As SyntaxNodeAnalysisContext)
            Dim interfaceDeclaration = DirectCast(context.Node, InterfaceStatementSyntax)
            Dim model = context.SemanticModel
            Dim name = interfaceDeclaration.Identifier.Text
            If (name.StartsWith("I", StringComparison.OrdinalIgnoreCase)) Then Exit Sub
            Dim errorMsg = String.Format(MessageFormat, MessageFormat)
            Dim diag = Diagnostic.Create(Rule, interfaceDeclaration.GetLocation(), errorMsg)
            context.ReportDiagnostic(diag)
        End Sub
    End Class
End Namespace